cmd/node/stats.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	pb "adaptodb/pkg/proto/proto"
+	"adaptodb/pkg/sm"
+
+	"github.com/lni/dragonboat/v3"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+
+type NodeStatsServer struct {
+	pb.UnimplementedNodeStatsServer
+	mu sync.Mutex
+	entries int64
+	successfulRequests int64
+	failedRequests int64
+	nh *dragonboat.NodeHost	
+	clusterID uint64
+	lastResetTime time.Time
+}
+
+func NewNodeStatsServer(nh *dragonboat.NodeHost, clusterID uint64) *NodeStatsServer {
+	return &NodeStatsServer{
+		nh: nh,
+		clusterID: clusterID,
+		lastResetTime: time.Now(),
+	}
+}
+
+// This method is called by the Balancer via gRPC to report the stats
+// of the current node. `successfulRequests` and `failedRequests` will
+// be reset after this call so each call to GetStats will report the
+// number of successful and failed requests since the last call - the
+// Balancer has the responsibility and the freedom to query this data
+// at regular intervals.
+func (s *NodeStatsServer) GetStats(ctx context.Context, req *pb.GetStatsRequest) (*pb.GetStatsResponse, error) {
+	// It might return a stale entry count, but it is good enough
+	// for metrics reporting purposes. We query the entry count
+	// using `nh` because we don't have access to the state machine
```

Comment:
What is nh?

Reply:
NodeHost. It's basically the "local Raft manager". We interact with Raft through `nh`